src/commands/dev/lib/dev-builder.ts

```diff
@@ -0,0 +1,296 @@
+
+import { globBuilderInputs } from './glob';
+import DevServer from './dev-server';
+import wait from '../../../util/output/wait';
+import IGNORED from '../../../util/ignored';
+import { NowError } from '../../../util/now-error';
+import { installBuilders, getBuilder, cacheDirPromise } from './builder-cache';
+import {
+  NowConfig,
+  BuildConfig,
+  BuilderInputs,
+  BuilderOutput,
+  BuilderOutputs,
+  BuiltLambda
+} from './types';
+
+/**
+ * Build project to statics & lambdas
+ */
+export async function buildUserProject(
+  nowJson: NowConfig,
+  devServer: DevServer
+): Promise<void> {
+  try {
+    devServer.setStatusBusy('Installing builders');
+    const builds = nowJson.builds || [];
+    await installBuilders(builds.map(build => build.use));
+
+    devServer.setStatusBusy('Building lambdas');
+    await executeBuilds(nowJson, devServer);
+
+    devServer.setStatusIdle();
+  } catch (err) {
+    devServer.setStatusIdle();
+    throw err;
+  }
+}
+
+export async function executeBuild(
+  nowJson: NowConfig,
+  devServer: DevServer,
+  asset: BuilderOutput
+): Promise<void> {
+  //console.error('rebuilding', { asset });
+  const { buildConfig, buildEntry } = asset;
+  if (!buildConfig || !buildEntry) {
+    throw new Error("Asset has not been built yet, can't rebuild");
+  }
+  const { cwd } = devServer;
+  const entrypoint = relative(cwd, buildEntry.fsPath);
+
+  const cacheDir = await cacheDirPromise;
+  const { dev, ino } = await stat(entrypoint);
+  const workPath = join(cacheDir, 'workPaths', String(dev + ino));
+  await mkdirp(workPath);
+
+  devServer.logDebug(`Building ${buildEntry.fsPath} (workPath = ${workPath})`);
+  const { builder } = buildConfig;
+  if (!builder) {
+    throw new Error('No builder');
+  }
+  const builderConfig = builder.config || {};
+  const files = await collectProjectFiles('**', cwd);
+  const config = buildConfig.config || {};
+  const output = await builder.build({
+    files,
+    entrypoint,
+    workPath,
+    config,
+    isDev: true
+  });
+
+  // enforce the lambda zip size soft watermark
+  const { maxLambdaSize = '5mb' } = { ...builderConfig, ...config };
+  let maxLambdaBytes: number;
+  if (typeof maxLambdaSize === 'string') {
+    maxLambdaBytes = bytes(maxLambdaSize);
+  } else {
+    maxLambdaBytes = maxLambdaSize;
+  }
+
+  for (const asset of Object.values(output)) {
+    if (asset.type === 'Lambda') {
+      const size = asset.zipBuffer.length;
+      if (size > maxLambdaBytes) {
+        throw new Error(
+          `The lambda function size (${bytes(
+            size
+          ).toLowerCase()}) exceeds the configured limit (${bytes(
+            maxLambdaBytes
+          ).toLowerCase()}). You may increase this by supplying \`maxLambdaSize\` to the build \`config\``
+        );
+      }
+    }
+
+    asset.buildConfig = buildConfig;
+    asset.buildEntry = buildEntry;
+  }
+
+  await Promise.all(
+    Object.entries(output).map(async e => {
```

Comment:
Is `e` an event? Is that a known convention or could we use a more descriptive name?